cmd/virt-prerunner/main.go

```diff
@@ -48,88 +48,36 @@ func main() {
-	}
-	cpuAffinity := ""
-	if len(vcpuToPCPU) > 0 {
-		cpuAffinity = fmt.Sprintf("[%s]", strings.Join(vcpuToPCPU, ","))
-	}
-	cloudHypervisorCmd = append(cloudHypervisorCmd, "--cpus", fmt.Sprintf("boot=%d,topology=%d:%d:%d:%d,affinity=%s",
-		vmConfig.Cpus.BootVcpus, vmConfig.Cpus.Topology.ThreadsPerCore, vmConfig.Cpus.Topology.CoresPerDie,
-		vmConfig.Cpus.Topology.DiesPerPackage, vmConfig.Cpus.Topology.Packages, cpuAffinity))
-
-	memoryArg := fmt.Sprintf("size=%d", vmConfig.Memory.Size)
-	if vmConfig.Memory.Shared {
-		memoryArg = memoryArg + ",shared=on"
-	}
-	if vmConfig.Memory.Hugepages {
-		memoryArg = memoryArg + ",hugepages=true"
-	}
-	cloudHypervisorCmd = append(cloudHypervisorCmd, "--memory", memoryArg)
-
-	if len(vmConfig.Disks) > 0 {
-		cloudHypervisorCmd = append(cloudHypervisorCmd, "--disk")
-		for _, disk := range vmConfig.Disks {
-			arg := fmt.Sprintf("id=%s,path=%s", disk.Id, disk.Path)
-			if disk.Readonly {
-				arg = arg + ",readonly=on"
-			}
-			if disk.Direct {
-				arg = arg + ",direct=on"
-			}
-			cloudHypervisorCmd = append(cloudHypervisorCmd, arg)
-		}
-	}
-
-	if len(vmConfig.Fs) > 0 {
-		cloudHypervisorCmd = append(cloudHypervisorCmd, "--fs")
-		for _, fs := range vmConfig.Fs {
-			arg := fmt.Sprintf("id=%s,socket=%s,tag=%s", fs.Id, fs.Socket, fs.Tag)
-			cloudHypervisorCmd = append(cloudHypervisorCmd, arg)
-		}
-	}
-
-	if len(vmConfig.Net) > 0 {
-		cloudHypervisorCmd = append(cloudHypervisorCmd, "--net")
-		for _, net := range vmConfig.Net {
-			if net.VhostUser {
-				cloudHypervisorCmd = append(cloudHypervisorCmd, fmt.Sprintf("id=%s,mac=%s,vhost_user=true,vhost_mode=server,socket=%s", net.Id, net.Mac, net.VhostSocket))
-			} else {
-				cloudHypervisorCmd = append(cloudHypervisorCmd, fmt.Sprintf("id=%s,mac=%s,tap=%s", net.Id, net.Mac, net.Tap))
-			}
-		}
+	chClient := getCloudHypervisorClient()
+	if err := chClient.VmCreate(context.TODO(), vmConfig); err != nil {
+		log.Fatalf("create VM: %s", err)
+		return
 	}
-
-	if len(vmConfig.Devices) > 0 {
-		cloudHypervisorCmd = append(cloudHypervisorCmd, "--device")
-		for _, device := range vmConfig.Devices {
-			cloudHypervisorCmd = append(cloudHypervisorCmd, fmt.Sprintf("id=%s,path=%s", device.Id, device.Path))
-		}
+	if err := chClient.VmBoot(context.TODO()); err != nil {
+		log.Fatalf("boot VM: %s", err)
+		return
 	}
-
-	fmt.Println(strings.Join(cloudHypervisorCmd, " "))
+	log.Println("booted VM")
 }
 
 func buildVMConfig(ctx context.Context, vm *virtv1alpha1.VirtualMachine) (*cloudhypervisor.VmConfig, error) {
 	vmConfig := cloudhypervisor.VmConfig{
+		Console: &cloudhypervisor.ConsoleConfig{
+			Mode: "Pty",
+		},
+		Serial: &cloudhypervisor.ConsoleConfig{
+			Mode: "Tty",
+		},
 		Payload: &cloudhypervisor.PayloadConfig{
 			Kernel: "/var/lib/cloud-hypervisor/hypervisor-fw",
 		},
 		Cpus: &cloudhypervisor.CpusConfig{
+			MaxVcpus:  int(vm.Spec.Instance.CPU.Sockets * vm.Spec.Instance.CPU.CoresPerSocket),
```

Comment:
This is required?

Reply:
yes, this field is required, as We don't support hotplug CPU now, so set it to this value.

Reply:
But it's not required in cmd args?

Reply:
according the [openapi](https://raw.githubusercontent.com/cloud-hypervisor/cloud-hypervisor/v26.0/vmm/src/api/openapi/cloud-hypervisor.yaml) `MaxVcpus` is required.